In [9]:
import os
os.environ["OPENAI_API_KEY"] = "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"


In [10]:
from openai import OpenAI
client = OpenAI()

Segment the orginal 40,000 terms into pieces small enough for chat-gpt 3.5 to read, less than 4000 tokens

In [12]:
def read_and_segment_file(filepath, max_length=3500):
    # Read the file
    with open(filepath, 'r') as file:
        content = file.read()

    # Split the content into terms
    terms = content.split(',')

    # Segment the terms
    segments = []
    current_segment = []
    current_length = 0

    for term in terms:
        term_length = len(term)
        if current_length + term_length > max_length:
            # Add the current segment to the segments list and start a new one
            segments.append(current_segment)
            current_segment = []
            current_length = 0

        # Add the term to the current segment
        current_segment.append(term)
        current_length += term_length + 1  # +1 for the comma

    # Add the last segment if it's not empty
    if current_segment:
        segments.append(current_segment)

    return segments

# Set the path to your file
file_path = "F:/VoLo/ukbiobank/participant_field_titles_no_tea.txt"

# Get the list of segmented terms
segmented_terms = read_and_segment_file(file_path)

# You can print the segments or process them as needed
#for segment in segmented_terms:
#    print(segment)


Add all responses to an object
Fead those one by one into Chat-gpt

In [17]:
from openai import OpenAI

client = OpenAI()

# Assuming segmented_terms is already defined and contains your data
responses = []  # List to store the responses

for segment_index, segment in enumerate(segmented_terms):
    # Join the terms in the current segment into a single string
    terms_string = ", ".join(segment)

    # Create the completion request for the current segment
    completion = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "system", "content": "You are an expert in Kidney disease."},
            {"role": "user", "content": f"Read from the list and tell me what terms might be related to kidney disease. Each term in this list is separated by a comma. Here is the list: {terms_string}"}
        ]
    )

    # Save the response for the current segment
    responses.append(completion.choices[0].message)

    # Optional: Print the response
    # print(f"Response for segment {segment_index + 1}: {completion.choices[0].message}")
    # print("\n" + "-"*50 + "\n")

# Now, 'responses' contains all the responses from the API.


In [31]:
len(responses)

432

In [30]:
print(responses[0])

ChatCompletionMessage(content='The terms related to kidney disease from the list are:\n- Blood sampling duration\n- Blood pressure\n- Hand grip dynamometer\n- Height measure\n- Manual scales\n- Waist circumference\n- Hip circumference\n- Standing height\n- Seated height', role='assistant', function_call=None, tool_calls=None)


In [32]:
processed_responses = []

for response in responses:
    # Extract the content from each response
    content = response.content

    # Remove the prefix
    prefix = "The terms related to kidney disease from the list are:\n"
    if content.startswith(prefix):
        content = content[len(prefix):]

    # Remove the suffix, if there's a consistent one to remove
    # e.g., content = content.replace("function_call=None, tool_calls=None)", "")

    # Add the processed content to the list
    processed_responses.append(content)

# 'processed_responses' now contains the cleaned content from each response


In [33]:
print(processed_responses[0])

- Blood sampling duration
- Blood pressure
- Hand grip dynamometer
- Height measure
- Manual scales
- Waist circumference
- Hip circumference
- Standing height
- Seated height


In [34]:
# Create a list to hold the cleaned terms
cleaned_terms = []

# Process each line in each response
for response in processed_responses:
    for line in response.split('\n'):
        # Remove the leading '- ' and any extra whitespace
        cleaned_line = line.strip('- ').strip()
        if cleaned_line:  # Ensure the line is not empty
            cleaned_terms.append(cleaned_line)

# Combine all cleaned terms into a single string
combined_terms = ', '.join(cleaned_terms)

# Now, 'combined_terms' is a single string with all terms separated by commas


In [36]:
# Write the final output
output_filename = "F:/VoLo/ukbiobank/kidney_terms4.txt"

# Open the file in write mode
with open(output_filename, 'w') as file:
    file.write(combined_terms)

# After this, 'combined_terms.txt' will contain all your terms in the specified format.


In [37]:
# Feed shorten list back in chat-gpt
def read_and_segment_file(filepath, max_length=3500):
    # Read the file
    with open(filepath, 'r') as file:
        content = file.read()

    # Split the content into terms
    terms = content.split(',')

    # Segment the terms
    segments = []
    current_segment = []
    current_length = 0

    for term in terms:
        term_length = len(term)
        if current_length + term_length > max_length:
            # Add the current segment to the segments list and start a new one
            segments.append(current_segment)
            current_segment = []
            current_length = 0

        # Add the term to the current segment
        current_segment.append(term)
        current_length += term_length + 1  # +1 for the comma

    # Add the last segment if it's not empty
    if current_segment:
        segments.append(current_segment)

    return segments

# Set the path to your file
file_path = "F:/VoLo/ukbiobank/kidney_terms4.txt"

# Get the list of segmented terms
segmented_terms2 = read_and_segment_file(file_path)

# You can print the segments or process them as needed
#for segment in segmented_terms:
#    print(segment)

In [62]:
from openai import OpenAI

client = OpenAI()

# Assuming segmented_terms is already defined and contains your data
responses_shorten = []  # List to store the responses

for segment_index, segment in enumerate(segmented_terms2):
    # Join the terms in the current segment into a single string
    terms_string = ", ".join(segment)

    # Create the completion request for the current segment
    completion = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "system", "content": "You are an expert in Kidney disease."},
            {"role": "user", "content": f"Read from the list and tell me what terms are strongly related to kidney disease. Each term in this list is separated by a comma. The output will be a list of terms seperated by commas. Absolutely, no commentary just the terms. If no related terms are found report None NA. Here is the list: {terms_string}"}
        ]
    )

    # Save the response for the current segment
    responses_shorten.append(completion.choices[0].message)

    # Optional: Print the response
    # print(f"Response for segment {segment_index + 1}: {completion.choices[0].message}")
    # print("\n" + "-"*50 + "\n")

# Now, 'responses' contains all the responses from the API.

In [63]:
print(responses_shorten[0])

ChatCompletionMessage(content='Blood pressure, Hand grip dynamometer, Waist circumference, Hip circumference, Standing height, Seated height', role='assistant', function_call=None, tool_calls=None)


In [71]:
processed_responses = []

for response in responses_shorten:
    # Extract the content from each response
    content = response.content

    # Remove the introductory sentence
    intro_sentence = "The terms related to kidney disease from the given list are:"
    content = content.replace(intro_sentence, "").strip()

    # Split the content into individual terms
    terms = content.split(', ')

    # Filter out 'None' and 'NA'
    filtered_terms = [term for term in terms if term.lower().strip() not in ["none", "na", "none na"]]

    # Rejoin the filtered terms and add to the list, ensuring it's not empty
    filtered_content = ', '.join(filtered_terms)
    if filtered_content:
        processed_responses.append(filtered_content)

# Combine all processed responses into a single string
combined_terms = ', '.join(processed_responses)

# Now you can write 'combined_terms' to a file or use it as needed


In [72]:
combined_terms

"Blood pressure, Hand grip dynamometer, Waist circumference, Hip circumference, Standing height, Seated height, Arterial stiffness, Signal-to-noise-ratio (SNR) of triplet (left), Triplet correct (left), Mean signal-to-noise ratio (SNR)\\, Time to press first digit (left) | Instance 0 | Array 1, Time to press first digit (left) | Instance 0 | Array 2, Time to press first digit (left) | Instance 0 | Array 3, Time to press first digit (left) | Instance 0 | Array 4, Time to press first digit (left) | Instance 0 | Array 5, Time to press first digit (left) | Instance 0 | Array 6, Time to press first digit (left) | Instance 0 | Array 7, Time to press first digit (left) | Instance 0 | Array 8, Time to press first digit (left) | Instance 0 | Array 9, Time to press first digit (left) | Instance 0 | Array 10, Time to press first digit (left) | Instance 0 | Array 11, Time to press first digit (left) | Instance 0 | Array 12, Time to press first digit (left) | Instance 0 | Array 13, Time to press fi

Write txt file with terms of interest.
From 40,000 down to 432 terms.

In [73]:
# Write the final output
output_filename = "F:/VoLo/ukbiobank/kidney_terms5_shorten.txt"

# Open the file in write mode
with open(output_filename, 'w') as file:
    file.write(combined_terms)

# After this, 'combined_terms.txt' will contain all your terms in the specified format.